pennylane/transforms/decompositions/clifford_plus_t/solovay_kitaev.py

```diff
@@ -0,0 +1,460 @@
+    s = ud @ qml.math.conj(qml.math.transpose(vwd))
+    sdg = vwd @ qml.math.conj(qml.math.transpose(ud))
+
+    # Get the required matrices V' and W'
+    v_hat = s @ v.matrix() @ sdg
+    w_hat = s @ w.matrix() @ sdg
+
+    return w_hat, v_hat
+
+
+def _approximate_umat(seqs, basic_approximations, KDTree=None):
+    """Approximates given GateSet using the TreeSet structure"""
+
+    if KDTree is None:
+        # use built-in min function for comparision - should be at least O(n^2)
+        def key(x):
+            return qml.math.linalg.norm(qml.math.subtract(x.so3_matrix, seqs.so3_matrix))
+
+        return min(basic_approximations, key=key)
+
+    # make use of the KD-Tree - should be at least O(log n)
+    seq_node = qml.math.array([qml.math.flatten(seqs.so3_matrix)])
+    _, index = KDTree.query(seq_node, workers=-1)
+
+    return basic_approximations[index[0]]
+
+
+def sk_approximate_set(basis_set=(), basis_depth=10, kd_tree=False):
+    r"""Builds approximate unitary set required for Solovay-Kitaev algorithm.
+
+    Args:
+        basis_set (list(str)): Basis set to be used for Solovay-Kitaev decomposition build using
+            following terms, ``['x', 'y', 'z', 'h', 't', 'tdg', 's', 'sdg']``. Default is ``["t", "tdg", "h"]``
+        basis_depth (int): Maximum expansion length of Clifford+T sequences in the approximation set. Default is `10`
+        kd_tree (bool): Return a KD-Tree corresponding to the gates in approximated set. Default is ``False``
+
+    Returns:
+        list or tuple(list, scipy.spatial.KDTree): A list of Clifford+T sequences that will be used for approximating
+        a matrix in the base case of recursive implementation of Solovay-Kitaev algorithm. If ``kd_tree`` argument
+        is ``True``, it also returns a KD-Tree built based on the list
+
+    .. seealso:: :func:`~.sk_decomposition` for performing Solovay-Kitaev decomposition.
+    """
+
+    if not basis_set:
+        basis_set = ["t", "tdg", "h"]
+
+    approximate_set = TreeSet(GateSet(), basis_set, ()).basic_approximation(depth=basis_depth)
+    if not kd_tree:
+        return approximate_set
+
+    gnodes = qml.math.array([qml.math.flatten(seq.so3_matrix) for seq in approximate_set])
+    return (approximate_set, sp.spatial.KDTree(gnodes))
+
+
+# pylint: disable=too-many-arguments
+def sk_decomposition(op, depth, basis_set=(), basis_depth=10, approximate_set=None, kd_tree=None):
+    r"""Approximate an arbitrary single-qubit gate in the Clifford+T basis using the `Solovay-Kitaev algorithm <https://arxiv.org/abs/quant-ph/0505030>`_.
+
+    This method implements a recursive Solovay-Kitaev decomposition that approximates any :math:`U \in \text{SU}(2)`
+    operation with :math:`\epsilon > 0` error. The error depends on the recursion ``depth``. In general, this
+    algorithm runs in :math:`O(\text{log}^{2.71}(1/\epsilon))` time and produces a decomposition with
+    :math:`O(\text{log}^{3.97}(1/\epsilon))` operations.
```

Comment:
should we clarify that this also means roughly the same order of T-gates specifically?

Reply:
Not sure! In the paper that I followed, they simply mentioned the complexity of gate operations and not T-gates, so I would prefer to leave it like that.